### PR TITLE
Allow for env vars to have spaces in some cases

### LINKED ISSUE #17781: Issue with ARG and EXPOSE when pass more than one ports
Issue when use external argument defined by ARG with EXPOSE and pass more than one ports:

**docker version:**
```
docker version
Client:
 Version:      1.9.0
 API version:  1.21
 Go version:   go1.4.3
 Git commit:   76d6bc9
 Built:        Tue Nov  3 19:20:09 UTC 2015
 OS/Arch:      darwin/amd64

Server:
 Version:      1.9.0
 API version:  1.21
 Go version:   go1.4.3
 Git commit:   76d6bc9
 Built:        Tue Nov  3 19:20:09 UTC 2015
 OS/Arch:      linux/amd64
```

**docker info:**
```
Containers: 12
Images: 64
Server Version: 1.9.0
Storage Driver: aufs
 Root Dir: /mnt/sda1/var/lib/docker/aufs
 Backing Filesystem: extfs
 Dirs: 88
 Dirperm1 Supported: true
Execution Driver: native-0.2
Logging Driver: json-file
Kernel Version: 4.1.12-boot2docker
Operating System: Boot2Docker 1.9.0 (TCL 6.4); master : 16e4a2a - Tue Nov  3 19:49:22 UTC 2015
CPUs: 1
Total Memory: 1.956 GiB
Name: default
ID: CDH7:HOHQ:F46Y:F4GN:LAFT:3VLC:42OL:UVZP:5E4N:PEWJ:HZHV:X4U4
Debug mode (server): true
 File Descriptors: 13
 Goroutines: 20
 System Time: 2015-11-07T08:56:56.849001113Z
 EventsListeners: 0
 Init SHA1: 
 Init Path: /usr/local/bin/docker
 Docker Root Dir: /mnt/sda1/var/lib/docker
Labels:
 provider=virtualbox
```

**name -a**:
```
Darwin pro.lan 15.0.0 Darwin Kernel Version 15.0.0: Sat Sep 19 15:53:46 PDT 2015; root:xnu-3247.10.11~1/RELEASE_X86_64 x86_64
```

To reproduce issue build image using:

Dockerfile:
```
FROM busybox
ARG EXPOSED_PORTS
RUN echo ${EXPOSED_PORTS}
EXPOSE ${EXPOSED_PORTS}
```
```
$ docker build --build-arg EXPOSED_PORTS="10001 10002" -t test .
```
Build log contain error message "Invalid containerPort: 10001 10002":

```
Sending build context to Docker daemon 2.048 kB
Step 1 : FROM busybox
 ---> c51f86c28340
Step 2 : ARG EXPOSED_PORTS
 ---> Using cache
 ---> 38b1bdfbf3aa
Step 3 : RUN echo ${EXPOSED_PORTS}
 ---> Running in 576441131c49
10001 10002
 ---> 2a4f56d378b6
Removing intermediate container 576441131c49
Step 4 : EXPOSE ${EXPOSED_PORTS}
Invalid containerPort: 10001 10002
pro:test himik$ cat Dockerfile 
FROM busybox
ARG EXPOSED_PORTS
RUN echo ${EXPOSED_PORTS}
EXPOSE ${EXPOSED_PORTS}
```

### PR DESCRIPTION
**Pull request by duglin**
Closes #17781

This allows for env vars in EXPOSE to be parsed for spaces so that each
"word" is then treated independently instead of as a single word/arg.

Signed-off-by: Doug Davis dug@us.ibm.com

### CODE REVIEW

File: builder/dockerfile/evaluator.go [hunk 1]
```diff
@@ -42,6 +42,19 @@ var replaceEnvAllowed = map[string]struct{}{
...
+// Certain commands are allowed to have their args split into more
+// words after env var replacements. Meaning:
+//   ENV foo="123 456"
+//   EXPOSE $foo
+// should result in the same thing as:
+//   EXPOSE 123 456
+// and not treat "123 456" as a single word.
+// Note that: EXPOSE "$foo" and EXPOSE $foo are not the same thing.
+// Quotes will cause it to still be treated as single word.
+var allowWordExpansion = map[string]struct{}{
```
Comment (calavera) [inline builder/dockerfile/evaluator.go]: please use `map[string]bool`

Comment (duglin) [inline builder/dockerfile/evaluator.go]: Done - but should https://github.com/docker/docker/pull/17789/files#diff-f1a9d9dacdc0756baf5e2bbf458d6e8eL32 be changed too?

Comment (calavera) [inline builder/dockerfile/evaluator.go]: @duglin I didn't want to add unrelated changes to the PR. But feel free to open a new PR and change that other map.

**Comment by MHBauer** [comment]
Nice tests. They seem comprehensive.

**Comment by vdemeester** [comment]
LGTM 🐶

**Comment by calavera** [comment]
LGTM

**Comment by thaJeztah** [comment]
ping @duglin was this a regression in 1.9 or only reported with 1.9?

**Comment by duglin** [comment]
@thaJeztah only reported in 1.9 - it wasn't a regression.

**Comment by thaJeztah** [comment]
@duglin thanks! checking if we didn't miss a PR for 1.9.1

**Pull request merged**